### PR TITLE
Faster _getSlideToSegment implementation

### PR DESCRIPTION
**Pull request by ransomr**
# Faster MergeTree._getSlideToSegment implementation

## Description

Previously _getSlideToSegment would walk all segments in the MergeTree. This replaces that implementation with a much more performant implementation by using rightExcursion and leftExcrusion to start the walk with the current segment in the tree.

## PR Checklist

-   [ ] I have updated the documentation accordingly.
-   [ ] I have added tests to cover my changes.
-   [X] All new and existing tests passed.
-   [X] My code follows the code style of this project.
-   [X] I ran the lint checks which produced no new errors nor warnings for my changes.
-   [X] I have checked to ensure there aren't other open Pull Requests for the same update/change.

## Does this introduce a breaking change?

-   [ ] Yes
-   [X] No

## Testing

Covered by `packages\dds\merge-tree\src\test\client.localReference.spec.ts`

### CODE REVIEW

File: packages/dds/merge-tree/src/mergeTree.ts [hunk 1]
```diff
@@ -1431,32 +1431,32 @@ export class MergeTree {
...
+        // Slide to the next farthest valid segment in the tree.
+        this.rightExcursion(segoff.segment, goFurtherToFindSlideToSegment);
+        if (slideToSegment) {
+            return { segment: slideToSegment, offset: 0 };
+        }
+        // If no such segment is found, slide to the last valid segment.
+        this.leftExcursion(segoff.segment, goFurtherToFindSlideToSegment);
+
+        // Workaround TypeScript issue (https://github.com/microsoft/TypeScript/issues/9998)
+        const identity = (seg: ISegment | undefined) => seg;
```
Comment (Abe27342) [inline packages/dds/merge-tree/src/mergeTree.ts:1451]: nit: I'd prefer casting; no need to add runtime behavior for ts deficiency. Not like this is a very hot code path though, so this is more the principle of the matter

**Review by Abe27342 — COMMENTED** [review]
Pretty sure with this change we need to convert `leftExcursion` to use `walkAllSegments` rather than `nodeMap` as was done for `rightExcursion`. Otherwise we fail to consider locally removed but not-yet-acked segments before the removed segment as valid slide targets. You'll need to make the immediately preceding segment (that's locally removed) be nested within a HierBlock though to get a repro.


edit: after further discussion with Ransom, this should be fine: nodeMapReverse doesn't do the same length checks that nodeMap does. We might want to clean that up later, but definitely unrelated to this PR :)

File: packages/dds/merge-tree/src/mergeTree.ts (hunk 1 above)
Comment (ransomr) [reply packages/dds/merge-tree/src/mergeTree.ts]: Fixed.

**Review by Abe27342 — APPROVED** [review]

**Comment by ransomr** [comment]
`leftExcursion`/`nodeMapReverse` don't do length checks, so they don't have the bug that `rightExcursion`/`nodeMap` had and the code should work as is.

**Pull request merged**